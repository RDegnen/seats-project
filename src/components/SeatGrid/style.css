.headerCol {
  width: 50px;
  height: 25px;
  margin: 0 5px;
  font-weight: bold;
  color: grey;
}

.headerRow {
  display: flex;
  justify-content: center;
}