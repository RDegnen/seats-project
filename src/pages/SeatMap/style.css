.selectedSeatRow {
  display: flex;
  justify-content: center;
}

.selectedSeatCol {
  margin: 10px 20px;
  font-weight: bold;
  color: grey;
  font-size: 30px;
}