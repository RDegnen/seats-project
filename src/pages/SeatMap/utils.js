// Group seats into objects based off seat properties
export const groupBy = (seats, prop) => {
  const returnObj = {}
  seats.forEach(seat => {
    if (returnObj[seat[prop]]) {
      returnObj[seat[prop]].push(seat)
    } else {
      returnObj[seat[prop]] = [seat]
    }
  })
  return returnObj
}
// Sort the seats in a row
export const sortRows = rows => {
  const sortedSeats = {...rows}
  for (const key in sortedSeats) {
    sortedSeats[key].sort((a, b) => a.seat > b.seat)
  }
  return sortedSeats
}
// Add ailes to all rows
export const addAisles = planeClass => {
  const alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
  Object.keys(planeClass).forEach(key => {
    const row = planeClass[key]
    for (let i = 0; i < row.length; i++) {
      if (row[i].seat !== alpha.charAt(i)) {
        row.splice(i, 0, { 
          row: parseInt(key),
          seat: alpha.charAt(i),
          aisle: true 
        })
      }
    }
  })
}
// Create the seat layout 
export const createLayout = res => {
  const classes = groupBy(res, 'class')
  const seats = {}
  for (const key in classes) {
    seats[key] = sortRows(groupBy(classes[key], 'row'))
  }
  Object.keys(seats).map(key => addAisles(seats[key]))
  return seats
}