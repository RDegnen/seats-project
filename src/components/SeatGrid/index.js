import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { Container, Row, Col } from 'reactstrap'

import SeatRow from '../SeatRow'
import { classes } from '../../constants'
import { seatPropType } from '../props'
import './style.css'

const ClassHeader = ({ row }) => (
  <Row className="headerRow">
    {row.map(({ seat, aisle }) => (
      <Col key={seat} className="headerCol">
        {aisle ? '' : seat}
      </Col>
    ))}
  </Row>
)

ClassHeader.propTypes = {
  row: PropTypes.arrayOf(seatPropType).isRequired,
}

const SeatGrid = ({ seats }) => {
  return (
    <Container>
      {[classes.FIRST, classes.BUSINESS, classes.ECONOMY].map(seatClass => {
        const classObj = seats[seatClass] || {}
        const firstRow = classObj[Object.keys(classObj)[0]] || []
        return (
          <Fragment key={seatClass}>
            <ClassHeader row={firstRow} />
            {Object.keys(classObj).map(rowKey => {
              const row = classObj[rowKey]
              return (
                <SeatRow key={`${rowKey}`} row={row} />
              )
            })}
          </Fragment>
        )})}
    </Container>
  )
}

SeatGrid.propTypes = {
  seats: PropTypes.object.isRequired
}

export default SeatGrid