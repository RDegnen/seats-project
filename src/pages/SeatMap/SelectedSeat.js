import React from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'reactstrap'

import './style.css'

const SelectedSeat = ({ seat, columns }) => (
  <Row className="selectedSeatRow">
    {columns.map(({ title, value }) => (
      <Col key={title} className="selectedSeatCol" md="3">
        {title}: {value(seat)}
      </Col>
    ))}
  </Row>
)

SelectedSeat.propTypes = {
  seat: PropTypes.object,
  columns: PropTypes.array,
}

export default SelectedSeat