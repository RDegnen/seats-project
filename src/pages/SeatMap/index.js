import React, { useEffect, useState } from 'react'

import SeatGrid from '../../components/SeatGrid'
import { createLayout } from './utils'

const seatId = (seatClass, row, seat) => `${seatClass}-${row}-${seat}`

const SeatMap = () => {
  const [seats, setSeats] = useState({})
  const [selectedSeat, setSelected] = useState(null)

  const getSeats = async () => {
    const res = 
      await fetch('https://s3.amazonaws.com/frontend-candidate-homework.lola.co/seats.json', {
          method: 'GET',
          mode: 'cors',
        }).then(res => res.json())
          .then(json => json)
          .catch(err => err)

    setSeats(createLayout(res))
  }

  useEffect(() => {
    getSeats()
  }, [])

  const selectSeat = item => {
    const { occupied, row, seat, class: seatClass } = item
    if (occupied) {
      return
    } else if (selectedSeat) {
      selectedSeat.selected = false
      if (seatId(
            selectedSeat.class, 
            selectedSeat.row, 
            selectedSeat.seat
          ) === seatId(seatClass, row, seat)) {
        setSelected(null)
        return
      }
    }
    item.selected = true
    setSelected(item)
  }

  return (
    <SeatGrid seats={seats} selectSeat={selectSeat}/>
  )
}

export default SeatMap