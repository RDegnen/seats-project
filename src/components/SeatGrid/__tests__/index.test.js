import React from 'react'
import { shallow } from 'enzyme'

import SeatGrid from '../'
import { createLayout } from '../../../pages/SeatMap/utils'

const testSeats = [
  {
    "seat": "B",
    "row": 3,
    "class": "First",
    "occupied": true
  },
  {
    "seat": "D",
    "row": 21,
    "class": "Economy",
    "occupied": false
  },
  {
    "seat": "C",
    "row": 12,
    "class": "Business",
    "occupied": true
  },
  {
    "seat": "A",
    "row": 3,
    "class": "First",
    "occupied": true
  },
  {
    "seat": "J",
    "row": 21,
    "class": "Economy",
    "occupied": false
  },
  {
    "seat": "F",
    "row": 12,
    "class": "Business",
    "occupied": false
  },
  {
    "seat": "A",
    "row": 21,
    "class": "Economy",
    "occupied": true
  },
]

const getComponent = props => shallow(
  <SeatGrid seats={createLayout(testSeats)} selectSeat={jest.fn()} />
)

describe('<SeatGrid />', () => {
  it('should match snapshot', () => {
    const component = getComponent()
    expect(component).toMatchSnapshot()
  })

  it('should add a header row to each class', () => {
    const component = getComponent()
    const headers = component.find('ClassHeader')
    const expectedFirstRow = [ 
      { seat: 'A', row: 3, class: 'First', occupied: true },
      { seat: 'B', row: 3, class: 'First', occupied: true },
    ]
    expect(headers.length).toBe(3)
    expect(headers.first().props().row).toEqual(expectedFirstRow)
    expect(
      headers
        .first()
        .dive()
        .find('Col')
        .first()
        .html()
        .toString())
        .toEqual('<div class="headerCol col">A</div>'
    )
  })
})