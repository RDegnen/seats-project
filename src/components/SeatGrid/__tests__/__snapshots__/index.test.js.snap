// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeatGrid /> should match snapshot 1`] = `
<Container
  tag="div"
>
  <ClassHeader
    row={
      Array [
        Object {
          "class": "First",
          "occupied": true,
          "row": 3,
          "seat": "A",
        },
        Object {
          "class": "First",
          "occupied": true,
          "row": 3,
          "seat": "B",
        },
      ]
    }
  />
  <SeatRow
    key="3"
    row={
      Array [
        Object {
          "class": "First",
          "occupied": true,
          "row": 3,
          "seat": "A",
        },
        Object {
          "class": "First",
          "occupied": true,
          "row": 3,
          "seat": "B",
        },
      ]
    }
    selectSeat={[MockFunction]}
  />
  <ClassHeader
    row={
      Array [
        Object {
          "aisle": true,
          "row": 12,
          "seat": "A",
        },
        Object {
          "aisle": true,
          "row": 12,
          "seat": "B",
        },
        Object {
          "class": "Business",
          "occupied": true,
          "row": 12,
          "seat": "C",
        },
        Object {
          "aisle": true,
          "row": 12,
          "seat": "D",
        },
        Object {
          "aisle": true,
          "row": 12,
          "seat": "E",
        },
        Object {
          "class": "Business",
          "occupied": false,
          "row": 12,
          "seat": "F",
        },
      ]
    }
  />
  <SeatRow
    key="12"
    row={
      Array [
        Object {
          "aisle": true,
          "row": 12,
          "seat": "A",
        },
        Object {
          "aisle": true,
          "row": 12,
          "seat": "B",
        },
        Object {
          "class": "Business",
          "occupied": true,
          "row": 12,
          "seat": "C",
        },
        Object {
          "aisle": true,
          "row": 12,
          "seat": "D",
        },
        Object {
          "aisle": true,
          "row": 12,
          "seat": "E",
        },
        Object {
          "class": "Business",
          "occupied": false,
          "row": 12,
          "seat": "F",
        },
      ]
    }
    selectSeat={[MockFunction]}
  />
  <ClassHeader
    row={
      Array [
        Object {
          "class": "Economy",
          "occupied": true,
          "row": 21,
          "seat": "A",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "B",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "C",
        },
        Object {
          "class": "Economy",
          "occupied": false,
          "row": 21,
          "seat": "D",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "E",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "F",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "G",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "H",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "I",
        },
        Object {
          "class": "Economy",
          "occupied": false,
          "row": 21,
          "seat": "J",
        },
      ]
    }
  />
  <SeatRow
    key="21"
    row={
      Array [
        Object {
          "class": "Economy",
          "occupied": true,
          "row": 21,
          "seat": "A",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "B",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "C",
        },
        Object {
          "class": "Economy",
          "occupied": false,
          "row": 21,
          "seat": "D",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "E",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "F",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "G",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "H",
        },
        Object {
          "aisle": true,
          "row": 21,
          "seat": "I",
        },
        Object {
          "class": "Economy",
          "occupied": false,
          "row": 21,
          "seat": "J",
        },
      ]
    }
    selectSeat={[MockFunction]}
  />
</Container>
`;
